File: src/1189_maximum_number_of_balloons.js
// problem link https://leetcode.com/problems/maximum-number-of-balloons
// time complexity O(n)
// space complexity O(n)

var maxNumberOfBalloons = function (text) {
    const balloonCach = {};
    const ballonSet = new Set(text.split(''));

    for (const char of text) {
        if (!ballonSet.has(char)) continue;

        const count = (balloonCach[char] ?? 0) + 1;
        balloonCach[char] = count;
    }

    let min = Math.min(
        balloonCach['b'],
        balloonCach['a'],
        balloonCach['n'],
        Math.floor(balloonCach['l'] / 2),
        Math.floor(balloonCach['o'] / 2),
    );

    return min ? min : 0;
};

module.exports = { maxNumberOfBalloons };

if (require.main === module) {
    const assert = require('assert');
    function ListNode(val = 0, next = null) { this.val = val; this.next = next; }
    function TreeNode(val = 0, left = null, right = null) { this.val = val; this.left = left; this.right = right; }
    function Node(val = 0, neighbors = []) { this.val = val; this.neighbors = neighbors; }
    function listNodeToString(node) {
        const values = [];
        let current = node;
        while (current !== null) {
            values.push(String(current.val));
            current = current.next;
        }
        return values.join('->');
    }
    function treeToString(node) {
        if (node === null) return 'None';
        return `${node.val}(${treeToString(node.left)},${treeToString(node.right)})`;
    }
    function sorted(value) { return value.slice().sort(); }
    assert.deepStrictEqual(maxNumberOfBalloons("nlaebolko"), 1);
    assert.deepStrictEqual(maxNumberOfBalloons("loonbalxballpoon"), 2);
    assert.deepStrictEqual(maxNumberOfBalloons("leetcode"), 0);
}

File: src/1189_maximum_number_of_balloons.py
# LeetCode 1189: Maximum Number Of Balloons
# Python version

def maxNumberOfBalloons(text: str) -> int:
    b = 0
    a = 0
    l = 0
    o = 0
    n = 0
    for ch in text:
        if ch == "b":
            b += 1
        elif ch == "a":
            a += 1
        elif ch == "l":
            l += 1
        elif ch == "o":
            o += 1
        elif ch == "n":
            n += 1

    res = min(b, a)
    res = min(res, l // 2)
    res = min(res, o // 2)
    res = min(res, n)
    return res



def main():
    assert maxNumberOfBalloons('nlaebolko') == 1
    assert maxNumberOfBalloons('loonbalxballpoon') == 2
    assert maxNumberOfBalloons('leetcode') == 0

if __name__ == "__main__":
    main()
